Source/Dafny/Resolver.cs

```diff
@@ -8461,75 +8478,21 @@ class GhostInterest_Visitor {
-              autoGhostIdExpr.Var.MakeGhost();
-            }
-          }
-
-          var gk = AssignStmt.LhsIsToGhost_Which(lhs);
-          if (gk == AssignStmt.NonGhostKind.IsGhost) {
-            s.IsGhost = true;
-            if (s.Rhs is TypeRhs) {
-              Error(s.Rhs.Tok, "'new' is not allowed in ghost contexts");
-            }
-          } else if (gk == AssignStmt.NonGhostKind.Variable && codeContext.IsGhost) {
-            // cool
-          } else if (mustBeErasable) {
-            Error(stmt, "Assignment to {0} is not allowed in this context (because this is a ghost method or because the statement is guarded by a specification-only expression)",
-              AssignStmt.NonGhostKind_To_String(gk));
-          } else {
-            if (gk == AssignStmt.NonGhostKind.Field) {
-              var mse = (MemberSelectExpr)lhs;
-              resolver.CheckIsCompilable(mse.Obj, codeContext);
-            } else if (gk == AssignStmt.NonGhostKind.ArrayElement) {
-              resolver.CheckIsCompilable(lhs, codeContext);
-            }
-
-            if (s.Rhs is ExprRhs) {
-              var rhs = (ExprRhs)s.Rhs;
-              if (!AssignStmt.LhsIsToGhost(lhs)) {
-                resolver.CheckIsCompilable(rhs.Expr, codeContext);
-              }
-            } else if (s.Rhs is HavocRhs) {
-              // cool
-            } else {
-              var rhs = (TypeRhs)s.Rhs;
-              if (rhs.ArrayDimensions != null) {
-                rhs.ArrayDimensions.ForEach(ee => resolver.CheckIsCompilable(ee, codeContext));
-                if (rhs.ElementInit != null) {
-                  resolver.CheckIsCompilable(rhs.ElementInit, codeContext);
-                }
-                if (rhs.InitDisplay != null) {
-                  rhs.InitDisplay.ForEach(ee => resolver.CheckIsCompilable(ee, codeContext));
-                }
-              }
-              if (rhs.InitCall != null) {
-                for (var i = 0; i < rhs.InitCall.Args.Count; i++) {
-                  if (!rhs.InitCall.Method.Ins[i].IsGhost) {
-                    resolver.CheckIsCompilable(rhs.InitCall.Args[i], codeContext);
-                  }
-                }
-              }
-            }
-          }
+          CheckAssignStmt(s, mustBeErasable, proofContext);
 
         } else if (stmt is CallStmt) {
           var s = (CallStmt)stmt;
           var callee = s.Method;
           Contract.Assert(callee != null);  // follows from the invariant of CallStmt
           s.IsGhost = callee.IsGhost;
-          // check in-parameters
-          if (mustBeErasable) {
+          if (proofContext != null && !callee.IsLemmaLike) {
+            Error(s, $"in {proofContext}, calls are allowed only to lemmas");
+          } else if (mustBeErasable) {
             if (!s.IsGhost) {
               Error(s, "only ghost methods can be called from this context");
             }
           } else {
+            // check in-parameters
```

Comment:
What's this comment doing here? Is this a TODO ?

Reply:
It describes the dozen code lines that follow. I moved this comment one line down, which makes that a little more apparent.